### PR TITLE
ci: publish crates directly

### PR DESCRIPTION
**Pull request by yihau**
#### Problem

- we don't raise errors when publishing crates
- we try to verify a crate even if it fails to publish
- the verification seems a little bit cumbersome

a related convo: https://discord.com/channels/428295358100013066/560503042458517505/1196572613523279942

#### Summary of Changes

1. return an error when the crate fails to publish
=> sometimes we receive reports about missing crates. I think we can detect these issues earlier and take appropriate actions (like re-publish or move to next tag)

2. use the check which is already in `cargo publish` instead of crate test
=> I guess we use crate-test for ensuring the successful publishing of the next crate 🤔 maybe we can just retry `cargo publish`. this command will tell us if anything goes wrong. (like dependences are not ready or building error)

these are big changes to our existing publish pipeline. let me know what you think. btw, I will only backport it to v1.17. I think our v1.16 will never pass this one atm🙈

### CODE REVIEW

File: ci/publish-crate.sh [hunk 1]
```diff
@@ -63,37 +63,27 @@ for Cargo_toml in $Cargo_tomls; do
...
-      ) && really_uploaded=1
-      if ((really_uploaded)); then
-        break;
+
+    numRetries=10
+    for ((i = 1; i <= numRetries; i++)); do
+      echo "Attempt ${i} of ${numRetries}"
+      # The rocksdb package does not build with the stock rust docker image so use
+      # the solana rust docker image
+      if ci/docker-run.sh "$rust_stable_docker_image" bash -exc "cd $crate; $cargoCommand"; then
```
[the remark this answers is not in the document]
Comment (t-nelson) [reply ci/publish-crate.sh:75]: we have `set -e` at ln2 though. pretty sure the script will bail on that

[the remark this answers is not in the document]
Comment (yihau) [reply ci/publish-crate.sh:75]: yeah. I think we can improve the error handling later. The retrying job behavior should be the same, the check is still like: 
1. skip when `publish = false` present in Cargo.toml
2. skip when it has already published 
3. try to publish

[the remark this answers is not in the document]
Comment (t-nelson) [reply ci/publish-crate.sh:75]: oh. (2) is what i was worried about. where do we check that?

[the remark this answers is not in the document]
Comment (yihau) [reply ci/publish-crate.sh:75]: here! https://github.com/solana-labs/solana/pull/34794/files#diff-ca91dcc53dd8d96f09b485711d45bc00450b3091cd00ab6ab98adf9213fdca3fL59-L62

**Review by t-nelson — APPROVED** [review]
:shipit:

**Pull request merged**